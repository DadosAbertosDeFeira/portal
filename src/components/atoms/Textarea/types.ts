import type { ComponentPropsWithoutRef, ComponentPropsWithRef } from "react";

export type TextareaProps = ComponentPropsWithRef<"textarea"> & {
  name: string;
  label: string;
  hideLabel?: boolean;
  errorText?: string;
  containerProps?: ComponentPropsWithoutRef<"label">;
  helperText?: string;
  filled?: boolean;
};
