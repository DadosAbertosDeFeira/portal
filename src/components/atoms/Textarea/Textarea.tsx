import { Box } from "atoms/Box";
import { Text } from "atoms/Text";
import { forwardRef, useMemo } from "react";
import { twMerge } from "tailwind-merge";

import type { TextareaProps } from "./types";

export const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>(
  function Textarea(
    {
      name,
      label,
      hideLabel = false,
      errorText,
      helperText,
      className,
      containerProps,
      filled = true,
      ...props
    },
    ref
  ) {
    const handleBottomText = useMemo(() => {
      if (errorText) {
        return (
          <Text
            id={`input-${name}--error-message`}
            className="text-sm text-alert"
          >
            {errorText}
          </Text>
        );
      }
      if (helperText) {
        return (
          <Text id={`input-${name}--describedby`} className="text-sm">
            {helperText}
          </Text>
        );
      }

      return null;
    }, [helperText, errorText]);

    return (
      <label
        {...containerProps}
        className={twMerge(
          "flex cursor-pointer flex-col flex-nowrap",
          containerProps?.className
        )}
        htmlFor={name}
      >
        <span className={hideLabel ? "sr-only" : ""}>{label}</span>
        <Box className="flex flex-col flex-nowrap gap-y-1">
          <textarea
            className={twMerge(
              "w-full rounded border border-gray-2 bg-white p-2 pl-4 font-medium text-gray-dark",
              className
            )}
            id={name}
            name={name}
            aria-errormessage={`textarea-${name}--error-message`}
            aria-describedby={`textarea-${name}--describedby`}
            placeholder={filled ? label : ""}
            {...props}
            ref={ref}
          />
          {handleBottomText}
        </Box>
      </label>
    );
  }
);
