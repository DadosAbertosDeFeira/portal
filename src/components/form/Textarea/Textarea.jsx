import React from 'react';
import PropTypes from 'prop-types';

function Textarea({ name, label, required }) {
  return (
    <textarea
      className="input-field h-20"
      aria-label={label}
      placeholder={label}
      name={name}
      id={name}
      required={required}
    />
  );
}

Textarea.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  required: PropTypes.bool,
};

Textarea.defaultProps = {
  required: false,
};

export default Textarea;
