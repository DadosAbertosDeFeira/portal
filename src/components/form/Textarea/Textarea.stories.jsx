/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import { useForm } from "react-hook-form";
import { Textarea } from "./Textarea";

export default {
  title: "components/Textarea",
  component: Textarea,
  argTypes: {
    control: {
      control: false,
    },
    name: {
      control: false,
    },
  },
};

export const Default = (args) => {
  const form = useForm();

  return <Textarea {...args} control={form.control} />;
};

export const Invalid = (args) => {
  const form = useForm();

  form.setError(args.name, {
    type: "required",
    message: "Esse valor é obrigatório.",
  });

  form.setValue(args.name, " ");

  return (
    <Textarea {...args} control={form.control} rules={{ required: true }} />
  );
};

Invalid.args = {
  name: "Email",
  label: "E-mail",
};
