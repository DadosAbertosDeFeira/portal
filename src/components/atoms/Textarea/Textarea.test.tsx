import { render } from "@testing-library/react";

import { Textarea, type TextareaProps } from "./Textarea";

describe("Textarea", () => {
  const makeSut = ({ ...props }: Partial<TextareaProps>) => {
    return render(<Textarea {...props} />);
  };

  it("renders correctly", () => {
    const { asFragment } = makeSut({});

    expect(asFragment).toMatchSnapshot();
  });
});
