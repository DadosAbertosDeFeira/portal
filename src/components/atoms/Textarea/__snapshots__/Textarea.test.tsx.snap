// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea renders correctly 1`] = `[Function]`;
